### ISSUE TITLE
Log probability on MSEDistribution is not executed via tfd.Independent

### ISSUE THREAD

**Issue by Corea** [google-research/planet#25]
I experienced that gradients were often vanished when I minimize reconstruction loss on other datasets. So I spent some time on debugging your model, and I found that [`log_prob` implementation on `MSEDistribution`](https://github.com/google-research/planet/blob/master/planet/tools/mse_distribution.py#L42) has not been used [when training loss is calculated](https://github.com/google-research/planet/blob/master/planet/training/utility.py#L205). 

Since your paper has the quote below, I guess this is some kind of mistake.
> Note that the log-likelihood under a Gaussian distribution with unit variance equals the mean squared error up to a constant.

Could you check this point?

**Comment by Corea** [comment]
I forgot to mention what should be changed. The name of `log_prob` function on `MSEDistribution` should be `_log_prob`.

**Comment by piojanu** [comment]
Also there is a bug in the function in line 45:
https://github.com/google-research/planet/blob/9cd9abed5b9a8831388f4d9da16e5604cfbd7c20/planet/tools/mse_distribution.py#L42-L47
`.ndims` should be inside the parenthesis `event_dims = [-(x + 1) for x in range(self.event_shape.ndims)]`. Although, I don't see the benefit of using this MSE dist. over regular Normal pdf. It was intended to be faster or what?

**Comment by danijar** [comment]
Thanks for pointing this out, this is good to know! Are you saying that `tfd.Independent` calls `_log_prob()` on the inner distribution instead of `log_prob()`? By the way, you can just use the Normal distribution from TensorFlow Probability instead, it works just fine.

**Comment by danijar** [comment]
Are you sure `MSEDistribution.log_prob()` is not called? I don't see how else it can compute a likelihood. To clarify, are you saying that `tfd.Independent` calls `_log_prob()` on the inner distribution instead of `log_prob()`? (@piojanu Thanks, that's definitely a typo but the code path isn't triggered.)

The intention of the `MSEDistribution` is to get more interpretable summaries that show MSE rather than log probs. But can just use the Normal distribution from TensorFlow Probability instead, I'm using it and it works just fine. I will probably update the code to make it the default at some point.

**Comment by Corea** [comment]
I think so. Now, the code has similar structure to the code below.

```python
dist = MSEDistribution(mean)
dist = tfd.Independent(distribution=dist)
dist.log_prob(output)
```

`tfd.Independent` doesn't have `log_prob` function, so [`log_prob` function of `tfd.Distribution`](https://github.com/tensorflow/probability/blob/v0.5.0/tensorflow_probability/python/distributions/distribution.py#L692-L716), the parent class of `tfd.Independent`, will be invoked. After following the codes there, [`_log_prob` function of `tfd.Independent`](https://github.com/tensorflow/probability/blob/v0.5.0/tensorflow_probability/python/distributions/independent.py#L195-L200) will be called. Finally, `_log_prob` function of `MSEDistribution` will be the destination. 

Nevertheless, I guess that this problem doesn't matter on training process.

**Comment by Corea** [comment]
I just found that the difference between package dependencies on `README.md` and `setup.py`. I tested planet on tensorflow 1.12 and tensorflow-probability 0.5 (which `setup.py` requires), while you might test this on 1.13.1 and 0.6 as `README.md` says. 

In [this commit](https://github.com/tensorflow/probability/commit/34e969a81a7bb96791104d8503b253e4f937c353), tensorflow-probability has changed their codes not to call `_log_prob` and to call `log_prob`. I think writing both functions (`_log_prob`, `log_prob`) is the best way.

**Comment by danijar** [comment]
Thanks for looking into this! Yes, the readme has the updated dependencies. Are you getting an error with the previous version of TFP? If the code runs, it must be calling the correct log_prob() implementation since there is no other one that it could call, as far as I can see.

**Comment by danijar** [comment]
Closing as I don't think is the case. If you still think there is a problem here, please comment.
